src/webgpu/web_platform/copyToTexture/image.spec.ts

```diff
@@ -0,0 +1,366 @@
+  `
+  )
+  .params(u =>
+    u
+      .combine('orientation', ['none', 'flipY'] as const)
+      .combine('srcDoFlipYDuringCopy', [true, false])
+      .combine('dstColorFormat', kValidTextureFormatsForCopyE2T)
+      .combine('dstPremultiplied', [true, false])
+      .beginSubcases()
+      .combine('width', [1, 2, 4, 15, 255, 256])
+      .combine('height', [1, 2, 4, 15, 255, 256])
+  )
+  .beforeAllSubcases(t => {
+    t.skipIfTextureFormatNotSupported(t.params.dstColorFormat);
+  })
+  .fn(async t => {
+    const {
+      width,
+      height,
+      orientation,
+      dstColorFormat,
+      dstPremultiplied,
+      srcDoFlipYDuringCopy,
+    } = t.params;
+
+    // CTS sometimes runs on worker threads, where document is not available.
+    // In this case, OffscreenCanvas can be used instead of <canvas>.
+    // But some browsers don't support OffscreenCanvas, and some don't
+    // support '2d' contexts on OffscreenCanvas.
+    // In this situation, the case will be skipped.
+    let imageCanvas: HTMLCanvasElement | OffscreenCanvas;
+    if (typeof document !== 'undefined') {
+      imageCanvas = document.createElement('canvas');
+      imageCanvas.width = width;
+      imageCanvas.height = height;
+    } else if (typeof OffscreenCanvas === 'undefined') {
+      t.skip('OffscreenCanvas is not supported');
+      return;
+    } else {
+      imageCanvas = new OffscreenCanvas(width, height);
+    }
+    const imageCanvasContext = imageCanvas.getContext('2d');
+    if (imageCanvasContext === null) {
+      t.skip('OffscreenCanvas "2d" context not available');
+      return;
+    }
+
+    // Generate non-transparent pixel data to avoid canvas
+    // different opt behaviour on putImageData()
+    // from browsers.
+    const texelViewSource = makeTestColorsTexelView({
+      testColors: kTestColorsOpaque,
+      format: 'rgba8unorm', // ImageData is always in rgba8unorm format.
+      width,
+      height,
+      flipY: false,
+      premultiplied: false,
+    });
+    // Generate correct expected values
+    const imageData = new ImageData(width, height);
+    texelViewSource.writeTextureData(imageData.data, {
+      bytesPerRow: width * 4,
+      rowsPerImage: height,
+      subrectOrigin: [0, 0],
+      subrectSize: { width, height },
+    });
+
+    // Use putImageData to prevent color space conversion.
+    imageCanvasContext.putImageData(imageData, 0, 0);
+
+    const blobFromCanvas = new Promise((resolve, reject) => {
+      if (typeof document !== 'undefined') {
+        (imageCanvas as HTMLCanvasElement).toBlob(blob => resolve(blob));
+      } else {
+        (imageCanvas as OffscreenCanvas).convertToBlob().then(resolve, reject);
+      }
+    });
+    const blob = (await blobFromCanvas) as Blob;
+    const url = URL.createObjectURL(blob);
+    const image = new Image(imageCanvas.width, imageCanvas.height);
+    image.src = url;
+    await raceWithRejectOnTimeout(image.decode(), 5000, 'load image timeout');
```

Comment:
image.decode() isn't appropriate here because that asks the image to be ready to add to the DOM, but that's not what we're going to do. Instead we should do roughly:

```ts
await raceWithRejectOnTimeout(new Promise(resolve => {
    image.onload = () => { resolve(); };
}), 5000, 'load image timeout');
```

even though that's annoyingly clunky.

Reply:
I thought onload was supposed to make it 'ready'. I was certain there was a way to do this without `.decode()`. I'll see if I can figure it out.

The spec does already have this check, see content timeline step 7 here:
https://gpuweb.github.io/gpuweb/#dom-gpuqueue-copyexternalimagetotexture
> Let usability be ? check the usability of the image argument(source).

Reply:
> When switching to `onload`, tests don't pass anymore as the image needs to be fully decoded to be extracted for canvas.

Ahh, I suspect this is because onload was set after .src. It has to be set before, so that it'll already be registered when loading starts.

Anyway, I thought about this some more and decided it's not worth switching decode() to onload. Even though decode() is supposed to be for adding to the DOM I think browsers in practice don't mind it for WebGPU/WebGL uploads. If it works as it looks like it should (even though the spec doesn't guarantee it), it'll still save blocking time inside copyExternalImageToTexture where it would otherwise have to synchronously decode.